### PR TITLE
Expose a test fixture helper class

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

## Summary

It's not always straightforward to write automated tests that leverage `ign-gazebo` from external projects. Internal to Gazebo, we have various helper classes in [test/helpers](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo5/test/helpers) that make it easier to write integration tests. The goal of this PR is to expose a public API that makes it easier to write tests similar to those.

The main class being added in this PR is the `TestFixture`. See the added example for how to use it in an external project together with GTest. Example usage from the classes's docs:

```
/// // Load a world with a fixture
/// ignition::gazebo::TestFixture fixture("path_to.sdf");
///
/// // Register callbacks, for example:
/// fixture.OnPostUpdate([&](const gazebo::UpdateInfo &,
///   const gazebo::EntityComponentManager &_ecm)
///   {
///     // Add expectations here
///   }
///
/// // Run the server
/// fixture.Server()->Run(true, 1000, false);
```

Besides `test/helpers`, a lot of the inspiration here is coming from @arjo129 's work on a separate project.

### Differences from test/helpers

While working on the public API I made some improvements over what we are currently using internally:

* Added new `Server::AddSystem` APIs that allow us to load a plugin created on the fly, instead of needing to load a shared library. (i.e. no more need for `libMockPlugin.so`, see the `HelperSystem` instead)
* The `TestFixture`, unlike `test/helpers/Relay`, instantiates a server automatically. This saves even more boilerplate in test code.

## Test it

Run the new standalone example. I made an effort to use classes like `World` and `Model` on the example test in order to make it as approachable as possible. (i.e. no `Each` calls)

## TODO

TODO before merging:

- [x] Add tests for `Server::AddSystem`
- [x] Add tests for `TestFixture`
- [x] I'd like to add a 2nd example test that is more realistic, sending commands and reading sensor data
- [x] General cleanup (docs, style, etc)

TODO in the future:

* ~~Tying into the simulation update loop is very powerful, but it can be inconvenient for doing one-time setup operations, such as getting a model's entity ID.~~ Done in #930 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: examples/standalone/custom_server/CMakeLists.txt
Comment (ahcorde) [inline examples/standalone/custom_server/CMakeLists.txt:2]: This is not part of this PR. Do you mind to remove it ?

File: examples/standalone/gtest_setup/gravity.sdf
Comment (ahcorde) [inline examples/standalone/gtest_setup/gravity.sdf:29]: minor style issue
```suggestion
```

File: include/ignition/gazebo/TestFixture.hh
Comment (ahcorde) [inline include/ignition/gazebo/TestFixture.hh:55]:
```suggestion
  public: TestFixture(const std::string &_path);
```

File: src/TestFixture.cc
Comment (ahcorde) [inline src/TestFixture.cc:2]:
```suggestion
 * Copyright (C) 2021 Open Source Robotics Foundation
```

File: src/TestFixture.cc
Comment (ahcorde) [inline src/TestFixture.cc:98]:
```suggestion
TestFixture::TestFixture(const std::string &_path)
```

**Review by ahcorde — DISMISSED** [review]

**Review by arjo129 — COMMENTED** [review]
Regarding the TODO I've opened #930. This will make it easier to perform tests at startup. Most of my comments are minor nitpicks.

File: examples/standalone/custom_server/CMakeLists.txt
Comment (chapulina) [reply examples/standalone/custom_server/CMakeLists.txt:2]: I noticed it while working on this PR. It could go on its own PR, but I thought it was small enough to put it here, this way we save some CI cycles

**Comment by chapulina** [comment]
This PR is ready for review. I recommend starting by reviewing https://github.com/ignitionrobotics/ign-gazebo/pull/935 and https://github.com/ignitionrobotics/ign-gazebo/pull/936 first, which are subsets of this PR and will make it shorter.

Once this is approved targeted at Fortress, I'm planning to retarget to Citadel, update as needed, merge, and forward-port up to Fortress.

**Comment by chapulina** [comment]
@mjcarroll , I merged from `main` and cleaned it up. It's ready for review :+1:

**Review by caguero — APPROVED** [review]
It looks good to me. The only comment I have is the example requires CMake 3.11 and Ubuntu Bionic ships with 3.10. We claim that Fortress is supported in Ubuntu Bionic. I think this is OK as it's one example and there are other alternative ways to install CMake.

**Comment by chapulina** [comment]
The PR was approved at 9963485 when targeting `main` (v6). I rebased onto `ign-gazebo3` and updated the API to work there on a05623d.

I'm going to merge into `ign-gazebo3` when CI comes back happy.

If anyone wants to use this with `main` before it's forward-ported, checkout 9963485.

**Pull request merged**
